pkg/gatewayserver/redis/registry.go

```diff
@@ -86,3 +86,32 @@ func (r *GatewayConnectionStatsRegistry) Get(ctx context.Context, ids *ttnpb.Gat
 	}
 	return result, nil
 }
+
+// BatchGet returns the connection stats for a batch of gateways.
+// NotFound errors indicating that the gateway is either not connected
+// or is connected to a different cluster, are ignored.
+func (r *GatewayConnectionStatsRegistry) BatchGet(ctx context.Context,
+	ids []*ttnpb.GatewayIdentifiers,
+) ([]*ttnpb.GatewayConnectionStatsEntry, error) {
+	ret := make([]*ttnpb.GatewayConnectionStatsEntry, 0)
+	if _, err := r.Redis.TxPipelined(ctx, func(p redis.Pipeliner) error {
+		for _, gtwIDs := range ids {
+			uid := unique.ID(ctx, gtwIDs)
+			result := &ttnpb.GatewayConnectionStats{}
+			if err := ttnredis.GetProto(ctx, r.Redis, r.key(uid)).ScanProto(result); err != nil {
```

Comment:
This won't use the pipeline, and the smoking gun is that nothing references `p`.

The way Redis pipelines work is as follows:
1. You call `{Tx}Pipelined` with a function that will _queue_ commands.
2. The function will be called with this `redis.Pipeliner` to which you add commands, but do not read responses from yet, as the commands have not been sent to the server.
3. The commands are then sent to the server in a batched manner, then the responses come back and `TxPipelined` returns.
4. You can now access the commands that you queued and read their result. 

The description above is what I meant when writing the issue description (sorry it wasn't detailed enough). You can see an example of this in the NS matching registry:
https://github.com/TheThingsNetwork/lorawan-stack/blob/f6e9d73bd6d6ccf69474dcc47f34c5a68c70dc44/pkg/networkserver/redis/registry.go#L539-L555
1. The commands are declared.
2. Inside the pipeline function we just assign them (but we don't read the result).
3. After the `TxPipelined` runs, we actually read the result.

With all of that being said, you can actually simplify your life massively using [`MGET`](https://redis.io/commands/mget/). You just have to build a slice of keys (so basically calling `r.key` on each unique ID), call `MGET` on them, then walk the result slice of results and unmarshal the protos using `redis.UnmarshalProto`. 

You can see an example of how this would work here: https://github.com/TheThingsIndustries/lorawan-stack/blob/afc07d9b357086064cfb46b26ae06cfaf6cdfbcc/pkg/redis/redis.tti.go#L48-L66 .

Reply:
Oh nice. Thanks for the detailed explanation. I'm definitely a noob with Redis.